src/Psalm/Internal/Type/TypeParser.php

```diff
@@ -721,16 +718,16 @@ private static function getTypeFromGenericTree(
                 throw new TypeParseTreeException('Union types are not allowed in value-of type');
             }
 
-            if (!$param_union_types[0] instanceof TClassConstant) {
+            if (!$param_union_types[0] instanceof TArray
```

Comment:
couldn't you reuse TKeyOfArray::isViableTemplateType ?

Reply:
Yes, I adjusted `isViableTemplateType` to handle unions and can now use it there for `$param_union_types`.